base_layer/core/src/mining/miner.rs

```diff
@@ -231,44 +231,51 @@ impl Miner {
                 true => task::spawn(self.mining()),
                 false => task::spawn(self.not_mining()),
             };
-            futures::select! {
-            msg = block_event.select_next_some() => {
-                match *msg {
-                    BlockEvent::Verified((_, ref result)) => {
-                        //Miner does not care if the chain reorg'ed or just added a new block. Both cases means a new chain tip, so it needs to restart.
-                    match *result {
-                        BlockAddResult::Ok | BlockAddResult::ChainReorg(_) => {
-                        stop_mining_flag.store(true, Ordering::Relaxed);
-                        start_mining = true;
+            // This flag will let the future select loop again if the miner has not been issued a shutdown command.
+            let mut wait_for_miner = false;
+            'inner: while (!wait_for_miner) {
+                futures::select! {
+                msg = block_event.select_next_some() => {
+                    match *msg {
+                        BlockEvent::Verified((_, ref result)) => {
+                            //Miner does not care if the chain reorg'ed or just added a new block. Both cases means a new chain tip, so it needs to restart.
+                        match *result {
+                            BlockAddResult::Ok | BlockAddResult::ChainReorg(_) => {
+                            stop_mining_flag.store(true, Ordering::Relaxed);
+                            start_mining = true;
+                            wait_for_miner = true;
```

Comment:
Why not just `break` out of a `loop` instead of using the `while` and another bool flag?